File: src/services/findings.ts
/**
 * Findings Service - API calls for bookmarks, saved visuals, and highlights
 */

import { API_BASE_URL } from './api';

const API_BASE = API_BASE_URL;

export interface Finding {
  id: string;
  notebook_id: string;
  type: 'visual' | 'answer' | 'highlight' | 'source' | 'note';
  title: string;
  created_at: string;
  updated_at: string;
  content: Record<string, unknown>;
  tags: string[];
  starred: boolean;
}

export interface FindingsStats {
  total: number;
  by_type: Record<string, number>;
  starred: number;
}

export const findingsService = {
  async createFinding(
    notebookId: string,
    type: Finding['type'],
    title: string,
    content: Record<string, unknown>,
    tags?: string[],
    starred?: boolean
  ): Promise<Finding> {
    const response = await fetch(`${API_BASE}/findings`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        notebook_id: notebookId,
        type,
        title,
        content,
        tags: tags || [],
        starred: starred || false,
      }),
    });
    if (!response.ok) throw new Error('Failed to create finding');
    return response.json();
  },

  async getFindings(
    notebookId: string,
    options?: {
      type?: Finding['type'];
      starred?: boolean;
      tag?: string;
      limit?: number;
      offset?: number;
    }
  ): Promise<{ findings: Finding[]; count: number }> {
    const params = new URLSearchParams();
    if (options?.type) params.append('type', options.type);
    if (options?.starred) params.append('starred', 'true');
    if (options?.tag) params.append('tag', options.tag);
    if (options?.limit) params.append('limit', options.limit.toString());
    if (options?.offset) params.append('offset', options.offset.toString());

    const response = await fetch(
      `${API_BASE}/findings/${notebookId}?${params.toString()}`
    );
    if (!response.ok) throw new Error('Failed to fetch findings');
    return response.json();
  },

  async getFinding(notebookId: string, findingId: string): Promise<Finding> {
    const response = await fetch(
      `${API_BASE}/findings/${notebookId}/${findingId}`
    );
    if (!response.ok) throw new Error('Finding not found');
    return response.json();
  },

  async updateFinding(
    notebookId: string,
    findingId: string,
    updates: {
      title?: string;
      tags?: string[];
      starred?: boolean;
      content?: Record<string, unknown>;
    }
  ): Promise<Finding> {
    const response = await fetch(
      `${API_BASE}/findings/${notebookId}/${findingId}`,
      {
        method: 'PATCH',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(updates),
      }
    );
    if (!response.ok) throw new Error('Failed to update finding');
    return response.json();
  },

  async deleteFinding(notebookId: string, findingId: string): Promise<void> {
    const response = await fetch(
      `${API_BASE}/findings/${notebookId}/${findingId}`,
      { method: 'DELETE' }
    );
    if (!response.ok) throw new Error('Failed to delete finding');
  },

  async getStats(notebookId: string): Promise<FindingsStats> {
    const response = await fetch(
      `${API_BASE}/findings/${notebookId}/stats/summary`
    );
    if (!response.ok) throw new Error('Failed to fetch findings stats');
    return response.json();
  },

  // Helper: Save a visual as a finding
  async saveVisual(
    notebookId: string,
    title: string,
    visualData: {
      type: 'svg' | 'mermaid';
      code: string;
      template_id?: string;
      source_content?: string;
    },
    tags?: string[]
  ): Promise<Finding> {
    return this.createFinding(notebookId, 'visual', title, visualData, tags);
  },

  // Helper: Save a chat answer as a finding
  async saveAnswer(
    notebookId: string,
    title: string,
    answerData: {
      question: string;
      answer: string;
      citations?: unknown[];
    },
    tags?: string[]
  ): Promise<Finding> {
    return this.createFinding(notebookId, 'answer', title, answerData, tags);
  },

  // Helper: Save a highlight/quote as a finding
  async saveHighlight(
    notebookId: string,
    title: string,
    highlightData: {
      text: string;
      source_id: string;
      source_name: string;
      page?: number;
    },
    tags?: string[]
  ): Promise<Finding> {
    return this.createFinding(notebookId, 'highlight', title, highlightData, tags);
  },
};

File: backend/api/findings.py
"""
Findings API - Endpoints for saving, retrieving, and managing user findings.

Part of the Canvas architecture - enables bookmarking visuals, highlights,
answers, and notes from research.
"""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional, List, Dict, Any

from services.findings_store import get_findings_store

router = APIRouter(prefix="/findings", tags=["findings"])


class CreateFindingRequest(BaseModel):
    notebook_id: str
    type: str  # 'visual' | 'answer' | 'highlight' | 'source' | 'note'
    title: str
    content: Dict[str, Any]
    tags: Optional[List[str]] = None
    starred: bool = False


class UpdateFindingRequest(BaseModel):
    title: Optional[str] = None
    tags: Optional[List[str]] = None
    starred: Optional[bool] = None
    content: Optional[Dict[str, Any]] = None


class FindingResponse(BaseModel):
    id: str
    notebook_id: str
    type: str
    title: str
    created_at: str
    updated_at: str
    content: Dict[str, Any]
    tags: List[str]
    starred: bool


@router.post("", response_model=FindingResponse)
async def create_finding(request: CreateFindingRequest):
    """Create a new finding (bookmark, saved visual, highlight, etc.)."""
    store = get_findings_store()
    
    # Validate type
    valid_types = ['visual', 'answer', 'highlight', 'source', 'note']
    if request.type not in valid_types:
        raise HTTPException(
            status_code=400,
            detail=f"Invalid finding type. Must be one of: {valid_types}"
        )
    
    finding = await store.create_finding(
        notebook_id=request.notebook_id,
        finding_type=request.type,
        title=request.title,
        content=request.content,
        tags=request.tags,
        starred=request.starred,
    )
    
    return FindingResponse(**finding.to_dict())


@router.get("/{notebook_id}/stats/summary")
async def get_findings_stats(notebook_id: str):
    """Get statistics about findings for a notebook.
    
    NOTE: This route must be defined BEFORE /{notebook_id}/{finding_id}
    to avoid route conflict in FastAPI.
    """
    store = get_findings_store()
    
    stats = await store.get_stats(notebook_id)
    return stats


@router.get("/{notebook_id}")
async def get_findings(
    notebook_id: str,
    type: Optional[str] = None,
    starred: bool = False,
    tag: Optional[str] = None,
    limit: int = 50,
    offset: int = 0,
):
    """Get findings for a notebook with optional filters."""
    store = get_findings_store()
    
    findings = await store.get_findings(
        notebook_id=notebook_id,
        type_filter=type,
        starred_only=starred,
        tag_filter=tag,
        limit=limit,
        offset=offset,
    )
    
    return {
        "findings": [f.to_dict() for f in findings],
        "count": len(findings),
    }


@router.get("/{notebook_id}/{finding_id}", response_model=FindingResponse)
async def get_finding(notebook_id: str, finding_id: str):
    """Get a specific finding."""
    store = get_findings_store()
    
    finding = await store.get_finding(notebook_id, finding_id)
    if not finding:
        raise HTTPException(status_code=404, detail="Finding not found")
    
    return FindingResponse(**finding.to_dict())


@router.patch("/{notebook_id}/{finding_id}", response_model=FindingResponse)
async def update_finding(
    notebook_id: str,
    finding_id: str,
    request: UpdateFindingRequest,
):
    """Update a finding's title, tags, starred status, or content."""
    store = get_findings_store()
    
    updates = {}
    if request.title is not None:
        updates['title'] = request.title
    if request.tags is not None:
        updates['tags'] = request.tags
    if request.starred is not None:
        updates['starred'] = request.starred
    if request.content is not None:
        updates['content'] = request.content
    
    if not updates:
        raise HTTPException(status_code=400, detail="No updates provided")
    
    finding = await store.update_finding(notebook_id, finding_id, updates)
    if not finding:
        raise HTTPException(status_code=404, detail="Finding not found")
    
    return FindingResponse(**finding.to_dict())


@router.delete("/{notebook_id}/{finding_id}")
async def delete_finding(notebook_id: str, finding_id: str):
    """Delete a finding."""
    store = get_findings_store()
    
    success = await store.delete_finding(notebook_id, finding_id)
    if not success:
        raise HTTPException(status_code=404, detail="Finding not found")
    
    return {"success": True, "message": "Finding deleted"}
